nautobot/extras/jobs.py

```diff
@@ -241,7 +240,7 @@ def properties_dict(cls):
     @classmethod
     def _get_vars(cls):
         vars = OrderedDict()
-        for name, attr in cls.__dict__.items():
+        for name, attr in inspect.getmembers(cls):
```

Comment:
This fixes a bug where you can't set variables on base classes but also changes the default sort order of fields in jobs. They no longer sort by order of declaration but by alphabetical sort due to the behavior of [`inspect.getmembers`](https://docs.python.org/3/library/inspect.html#inspect.getmembers). Should this be considered a breaking change?

Reply:
Not sure if a breaking change, but it is unfortunate. Users will have to get in the habit of explicitly declaring a [field_order](https://nautobot.readthedocs.io/en/stable/additional-features/jobs/#field_order) on all Jobs if we can't find a better option here. (If we do go ahead with this change then we at least need to update those docs...)

As an alternative, maybe we could do something like iterating over all classes in `reversed(inspect.getmro(cls))` and iterating over the `__dict__` of each to look for variables?

Reply:
We could copy [`inspect.getmembers`](https://github.com/python/cpython/blob/915fb4f4418de5150121ed16ed88c2e11f60283b/Lib/inspect.py#L444) and remove the last line that sorts the results.

Reply:
Can we just iterate over the dictionary still, something like 
```python
        for name in cls.__dict__.keys():
            attr = [i[1] for i in inspect.getmembers(cls) if i[0] == name][0]
            if name not in vars and issubclass(attr.__class__, ScriptVariable):
                vars[name] = attr
```

Reply:
That fixes the sort issue but doesn't work if an attribute is only defined on a base class. We want to be able to define `object_change = ObjectVar(model=ObjectChange)` on the parent class and have it pass through to any jobs that subclass that.

Reply:
What about:
```
        for name, attr in inspect.getmembers(cls):
            attr_map[name] = attr
        for name in cls.__dict__.keys():
            if name not in vars and issubclass(attr.__class__, ScriptVariable):
                vars[name] = attr
                del attr_map[name]
        for name, attr in attr_map.items():
            if name not in vars and issubclass(attr.__class__, ScriptVariable):
                vars[name] = attr
```

Reply:
I iterated on your idea and came up with this. This puts the attributes in the order they were defined in the code and also puts attributes defined in the parent class before subclass attributes. Since we're using `getattr(cls, name)` we always get the correct attribute.

```py
    @classmethod
    def _get_vars(cls):
        vars = OrderedDict()
        base_classes = reversed(inspect.getmro(cls))
        attr_names = [name for base in base_classes for name in base.__dict__.keys()]
        for name in attr_names:
            attr_class = getattr(cls, name, None).__class__
            if name not in vars and issubclass(attr_class, ScriptVariable):
                vars[name] = getattr(cls, name)

        return vars
```

Reply:
What's the reason to re-introduce `OrderedDict`?

1. https://github.com/nautobot/nautobot/issues/1746
2. AFAIK the native dictionary in 3.7+ has all the same features as OrderedDict.

Reply:
This was carried over from the existing method but I can change it to dict as part of this change.